### ISSUE TITLE
--rules-stack with insert/overstrike rules doubles the word

### ISSUE THREAD

**Issue by solardiz** [openwall/john#4159]
This looks weird to me:
```
[solar@super run]$ cat w 
word
[solar@super run]$ ./john -w=w -ru=i1 -rules-stack=i1 -stdo | head
Using default input encoding: UTF-8
Press 'q' or Ctrl-C to abort, almost any other key for status
 wo rdword
!wo rdword
"wo rdword
#wo rdword
$wo rdword
%wo rdword
&wo rdword
'wo rdword
(wo rdword
)wo rdword
[solar@super run]$ ./john -w=w -ru=o1 -rules-stack=o1 -stdo | head
Using default input encoding: UTF-8
Press 'q' or Ctrl-C to abort, almost any other key for status
 or word
!or word
"or word
#or word
$or word
%or word
&or word
'or word
(or word
)or word
```

(This test is on non-latest tree, but I think we made no relevant changes recently?)

Edit: My actual use case for this was to mix inserts and overstrikes. Trying that in the way above also doubles the word.

**Comment by magnumripper** [comment]
I take it you only see problems when using --rules-stack, right?

**Comment by solardiz** [comment]
Yes, the issue is specifically with `--rules-stack`.

**Comment by AlekseyCherepanov** [comment]
It should be #4107, that's fixed already. #4107 is a minimal test case for failing i1+i1 or something like that. I did not check i1+i1 / o1+o1 after the fix though.

**Comment by magnumripper** [comment]
Confirmed, it was fixed in 6b14c5a.
